src/PoolManager.sol

```diff
@@ -207,34 +218,71 @@ contract PoolManager is IPoolManager, Fees, NoDelegateCall, ERC6909Claims {
         PoolId id = key.toId();
         _checkPoolInitialized(id);
 
-        key.hooks.beforeSwap(key, params, hookData);
-
-        uint256 feeForProtocol;
-        uint24 swapFee;
-        Pool.SwapState memory state;
-        (delta, feeForProtocol, swapFee, state) = pools[id].swap(
-            Pool.SwapParams({
-                tickSpacing: key.tickSpacing,
-                zeroForOne: params.zeroForOne,
-                amountSpecified: params.amountSpecified,
-                sqrtPriceLimitX96: params.sqrtPriceLimitX96
-            })
+        // The hook's deltas are from the point of view of the hook. Positive: the hook took money, negative: the hook sent money to the pool
+        // TODO consider adding hookDeltaInUnspecified to beforeSwap hook too
+        int128 hookDeltaInSpecified = key.hooks.beforeSwap(key, params, hookData);
+
+        bool exactInput = params.amountSpecified > 0;
+        params.amountSpecified -= hookDeltaInSpecified;
+        require(exactInput ? params.amountSpecified > 0 : params.amountSpecified < 0);
+
+        // ExactIn (amountSpecified is positive)
+        // In the case where the hook contributed x, x is given to the user now. After the swap amountSpecified + x input are taken from the user, meaning the user pays amountSpecified overall.
```

Comment:
I find these comments kinda confusing. When you say " After the swap amountSpecified + x input are taken from the user, meaning the user pays amountSpecified overall" it doesn't sound like the user only pays amount specified overall it sounds like +x is also being taken from the user...

Same in the case where you talk about hook taking x.

Maybe something like:
In the case where a hook returns a positive delta, the hook is requesting an amount to be drawn from the pool manager. Thus the hook delta is subtracted from the amountSpecified. Note that only the original amountSpecified is sent in from the user.

In the case where a hook returns a negative delta, the hook is requesting a debt to the pool manager (must owe the pool manager). Thus the hook delta ends up being an additive to the amountSpecified by the user. Again, note that only the original amountSpecified is sent in from the user.

These comments are super long though so maybe they deserve a separate natspec for them outside the function cause it ends up breaking up the code and gets harder to read sometimes. But thought I would just drop this here to make sure a) I'm thinking about it how you are and b) to note that the above ones are kinda confusing to me so worth rethinking the language

Reply:
yeah i do think theyre too long for this function.
maybe we're best to take them out entirely and then of course make clear docs about how custom accounting works elsewhere. Maybe a section of the readme (or elsewhere) on how hook returned deltas work.